scripts/prepare_sector_network.py

```diff
@@ -946,6 +951,87 @@ def add_storage(n, costs):
+
+        cols = [
+            "bus0",
+            "bus1",
+            "is_bothDirection",
+            "pipe_capacity_MW",
+            "id",
+            "length_km"
+        ]
+        gas_pipes = pd.read_csv(snakemake.input.clustered_gas_network)[cols]
+        index = "Gas pipeline " + gas_pipes["bus0"] + " -> " + gas_pipes["bus1"]
+        index[gas_pipes.is_bothDirection==1] = index[gas_pipes.is_bothDirection==1].str.replace("->", "<->")
+        gas_pipes.index = index
+
+        # group parallel pipes together
+        strategies = {
+            'bus0': 'first',
+            'bus1': 'first',
+            'is_bothDirection': 'first',
+            "pipe_capacity_MW": 'sum',
+            "length_km": 'sum',
+            'id': ' '.join,
+        }
+        num_parallel = gas_pipes.groupby(gas_pipes.index).count().iloc[:,0]
+        gas_pipes = gas_pipes.groupby(gas_pipes.index).agg(strategies)
+
+        both_direction_i = gas_pipes[gas_pipes.is_bothDirection==1].index
+        # TODO capital cost from DEA energy transport data sheet 102 6
+        # lifetime 50 a, investment ~1 EUR/MW/km, FOM 0.12 EUR/MW/km/year, discountrate 7%
+        pipe_fixed_cost = 0.19   # EUR/MW/km/a (annualised investment costs)
+        gas_pipes["p_min_pu"] = 0.
+        gas_pipes.loc[both_direction_i, "p_min_pu"] = -1
+
+        n.madd("Link",
+            gas_pipes.index,
+            bus0=gas_pipes.bus0 + " gas",
+            bus1=gas_pipes.bus1 + " gas",
+            p_min_pu=gas_pipes.p_min_pu,
+            p_nom=gas_pipes.pipe_capacity_MW,
+            p_nom_extendable=False,
+            length=gas_pipes.length_km,
+            capital_cost=gas_pipes.length_km * pipe_fixed_cost,
+            type=num_parallel,
+            tags=gas_pipes.id,
+            carrier="Gas pipeline",
+            lifetime=50
+        )
+        
+        # remove fossil generators at all connected nodes
+        # TODO what should be assumed here? rather located at LNG terminals?
+        missing = nodes.difference(pd.concat([gas_pipes.bus0, gas_pipes.bus1]).unique())
+        remove_i = n.generators[(n.generators.carrier=="gas")
+                              & (~n.generators.bus.str.replace(" gas","").isin(missing))].index
+        n.generators.drop(remove_i, inplace=True)
+
+        # retroftting existing CH4 pipes to H2 pipes
+        if options["H2_retrofit"]:
+
+            gas_pipe_i = n.links[n.links.carrier=="Gas pipeline"].index
+            n.links.loc[gas_pipe_i, "p_nom_extendable"] = True
+            h2_pipes = gas_pipes.rename(index=lambda x:
+                                        x.replace("Gas pipeline", "H2 pipeline retrofitted"))
+
+            n.madd("Link",
+                h2_pipes.index,
+                bus0=h2_pipes.bus0 + " H2",
+                bus1=h2_pipes.bus1 + " H2",
+                p_min_pu=-1.,  # allow that H2 pipelines can be used in other direction
+                p_nom_max=h2_pipes.pipe_capacity_MW,
```

Comment:
Doesn't that need a conversion from CH4 capacity to H2 capacity roughly by factor 3?

Reply:
Ah, not quite. Would this serve as a good approximation?

> Hydrogen has a lower energy density than natural gas: at the same pressure, a cubic meter of hydrogen only contains 1/3 of the energy of a cubic meter of natural gas. However, this does not mean that three times as many pipelines are required to transport the same amount of energy. The volume flow of hydrogen can be higher than for natural gas, bringing the maximum energy capacity of a hydrogen pipeline to a value of up to 80% of the energy capacity it has when transporting natural gas. 

https://gasforclimate2050.eu/wp-content/uploads/2020/07/2020_European-Hydrogen-Backbone_Report.pdf, p. 10

```suggestion
                p_nom_max=h2_pipes.pipe_capacity_MW * 0.8,
```

Reply:
I think we should move the factor of 80% into the `config.yaml ` to be more transparent. It can be more economical to operate with a lower pipe capacity than 80%. The pipe capacity of the retrofitted pipelines is also restricted by the [add_pipe_retrofit_constraint](https://github.com/PyPSA/pypsa-eur-sec/blob/81b009e0623f149c2c7695bd83ac56e43d5ae9fd/scripts/solve_network.py#L154)